File: public/dsa/linked_lists/intersection_of_two_linked_lists/approach/intersection_of_two_linked_lists.cpp
#include <bits/stdc++.h>
using namespace std;

// defs-start
struct ListNode {
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};
// defs-end

ListNode* getIntersectionNode(ListNode* A, ListNode* B) {
    // add your logic here
    auto tempA = A;
    auto tempB = B;
    while (tempA != tempB) {
        tempA = tempA ? tempA->next : B;
        tempB = tempB ? tempB->next : A;
    }
    return tempA;
}

// tests-start
int main() {
    ListNode* C = new ListNode(1, new ListNode(2));
    ListNode* A = new ListNode(11, new ListNode(12, new ListNode(13, C)));
    ListNode* B = new ListNode(21, new ListNode(22, C));

    auto temp = getIntersectionNode(A, B);
    if (temp) {
        cout << temp->data << endl;
    } else {
        cout << -1 << endl;
    }
    return 0;
}
// tests-end

/*output
1
*/

File: public/dsa/linked_lists/intersection_of_two_linked_lists/approach/intersection_of_two_linked_lists.py
from typing import Optional


# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def getIntersectionNode(self, A: ListNode, B: ListNode) -> Optional[ListNode]:
        tempA, tempB = A, B
        while tempA != tempB:
            tempA = tempA.next if tempA else B
            tempB = tempB.next if tempB else A
        return tempA


# tests-start
C = ListNode(1, ListNode(2))
A = ListNode(11, ListNode(12, ListNode(13, C)))
B = ListNode(21, ListNode(22, C))
sol = Solution()
temp = sol.getIntersectionNode(A, B)
if temp:
    print(temp.data)
else:
    print(-1)
# tests-end

'''output
1
'''
